marlowe-playground-client/src/Halogen/Monaco.purs

```diff
@@ -0,0 +1,130 @@
+        , handleQuery
+        , initialize: Just Init
+        , receive: const Nothing
+        , finalize: Nothing
+        }
+    }
+
+render :: forall p i. State -> HTML p i
+render state =
+  div
+    [ ref $ H.RefLabel "monacoEditor"
+    , class_ $ H.ClassName "monaco-editor-container"
+    ]
+    []
+
+handleAction :: forall slots m. MonadAff m => MonadEffect m => Settings m -> Action -> HalogenM State Action slots Message m Unit
+handleAction settings Init = do
+  monaco <- liftEffect Monaco.getMonaco
+  maybeElement <- H.getHTMLElementRef settings.refLabel
+  case maybeElement of
+    Just element -> do
+      let
+        languageId = view Monaco._id settings.languageExtensionPoint
+      liftEffect $ Monaco.registerLanguage monaco settings.languageExtensionPoint
+      liftEffect $ Monaco.defineTheme monaco settings.theme
+      liftEffect $ Monaco.setTokensProvider monaco languageId settings.tokensProvider
+      liftEffect $ Monaco.registerCompletionItemProvider monaco languageId settings.completionItemProvider
+      liftEffect $ Monaco.registerCodeActionProvider monaco languageId settings.codeActionProvider
+      liftEffect $ Monaco.registerDocumentFormattingEditProvider monaco languageId settings.documentFormattingEditProvider
+      editor <- liftEffect $ Monaco.create monaco element languageId settings.theme.name
+      void $ H.modify (const { editor: Just editor })
+      void $ H.subscribe $ effectEventSource (changeContentHandler monaco editor)
+      H.lift $ settings.setup editor
+    Nothing -> pure unit
+  where
+  changeContentHandler monaco editor (Emitter emitter) = do
+    Monaco.onDidChangeContent editor
+      ( \_ -> do
+          model <- Monaco.getModel editor
+          Monaco.setModelMarkers monaco model settings.owner settings.getModelMarkers
+          markers <- Monaco.getModelMarkers monaco model
+          emitter $ Left $ HandleChange (Monaco.getValue model) markers
+      )
+    pure $ Finalizer $ pure unit
+
+handleAction _ (HandleChange contents markers) = do
+  H.raise $ TextChanged contents
+  H.raise $ MarkersChanged markers
+
+handleQuery :: forall a input m. MonadEffect m => Query a -> HalogenM State Action input Message m (Maybe a)
+handleQuery (SetText text next) = do
+  H.gets _.editor
+    >>= traverse_ \editor -> do
+        model <- liftEffect $ Monaco.getModel editor
+        liftEffect $ Monaco.setValue model text
+  pure $ Just next
+
+handleQuery (GetText f) = do
+  H.gets _.editor
+    >>= traverse
+        ( \editor -> do
+            model <- liftEffect $ Monaco.getModel editor
+            let
+              s = Monaco.getValue model
+            pure $ f s
+        )
+
+handleQuery (SetPosition position next) = do
+  H.gets _.editor
+    >>= traverse_ \editor -> liftEffect $ Monaco.setPosition editor position
+  pure $ Just next
+
+handleQuery (Resize next) = do
+  H.gets _.editor >>= traverse_ \editor -> liftEffect $ Monaco.layout editor
```

Comment:
Do you think it's worth defining:

``` purescript
withEditor action = H.gets _.editor >>= traverse action
```

...such that:

``` purescript
handleQuery (Resize next) = withEditor (liftEffect <<< Monaco.layout)
```

...and so on? Might even want to roll the `liftEffect` inside...

Reply:
Done, although not rolling the `liftEffect` in because this would make it tricky to do other things